File: android/src/main/java/expo/modules/motionblurview/models/BlurCanvas.kt
package expo.modules.motionblurview.models

import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.roundToInt

internal data class BlurCanvas(val bounds: Bounds, val frame: Bounds, val scale: Double) {
  val contentOriginX: Int
    get() = ((bounds.left - frame.left) * scale).roundToInt()

  val contentOriginY: Int
    get() = ((bounds.top - frame.top) * scale).roundToInt()

  val pixelWidth: Int
    get() = (frame.width * scale).roundToInt()

  val pixelHeight: Int
    get() = (frame.height * scale).roundToInt()

  val isRenderable: Boolean
    get() = bounds.width > 0 && bounds.height > 0 &&
      pixelWidth <= MAX_PIXEL_DIMENSION && pixelHeight <= MAX_PIXEL_DIMENSION

  companion object {
    const val MAX_PIXEL_DIMENSION = 8192

    fun make(bounds: Bounds, inkRect: Bounds, intensity: Double, scale: Double, streakReach: Double): BlurCanvas {
      val streakRoom = ceil(max(intensity, 0.0) * scale * streakReach) + 2
      fun margin(ink: Double): Double = (ceil(max(ink, 0.0) * scale) + streakRoom) / scale

      val left = margin(bounds.left - inkRect.left)
      val top = margin(bounds.top - inkRect.top)
      val right = margin(inkRect.right - bounds.right)
      val bottom = margin(inkRect.bottom - bounds.bottom)
      return BlurCanvas(
        bounds = bounds,
        frame = Bounds(bounds.left - left, bounds.top - top, bounds.right + right, bounds.bottom + bottom),
        scale = scale
      )
    }
  }
}

File: ios/Models/BlurCanvas.swift
//
//  BlurCanvas.swift
//  Pods
//
//  Created by rit3zh CX on 9/23/26.
//

import CoreGraphics

struct BlurCanvas: Equatable {
  static let maxPixelDimension = 8192
  let bounds: CGRect
  let frame: CGRect
  let scale: CGFloat
  init(bounds: CGRect, inkRect: CGRect, intensity: CGFloat, scale: CGFloat) {
    self.bounds = bounds
    self.scale = scale
    let streakRoom = ceil(max(intensity, 0) * scale) + 2
    func margin(_ ink: CGFloat) -> CGFloat {
      (ceil(max(ink, 0) * scale) + streakRoom) / scale
    }
    let left = margin(bounds.minX - inkRect.minX)
    let top = margin(bounds.minY - inkRect.minY)
    let right = margin(inkRect.maxX - bounds.maxX)
    let bottom = margin(inkRect.maxY - bounds.maxY)
    frame = CGRect(
      x: bounds.minX - left,
      y: bounds.minY - top,
      width: bounds.width + left + right,
      height: bounds.height + top + bottom
    )
  }
  var contentOrigin: CGPoint {
    CGPoint(x: bounds.minX - frame.minX, y: bounds.minY - frame.minY)
  }
  var pixelWidth: Int {
    Int((frame.width * scale).rounded())
  }
  var pixelHeight: Int {
    Int((frame.height * scale).rounded())
  }
  var drawableSize: CGSize {
    CGSize(width: pixelWidth, height: pixelHeight)
  }
  var isRenderable: Bool {
    bounds.width > 0 && bounds.height > 0
      && pixelWidth <= Self.maxPixelDimension && pixelHeight <= Self.maxPixelDimension
  }
}
